Support for Pushed Authorization (PAR) in OIDC Handler

josephdecock opened the pull request:
# Support for Pushed Authorization Requests

Adds support for Pushed Authorization Requests (RFC 9126) to the OIDC authentication handler.

josephdecock committed:
Initial implementation of PAR in OIDC Handler

Adds support for pushed authorization requests (RFC 9126) to the OpenIdConnect authentication handler.

josephdecock committed:
Fix API declaration

josephdecock commented:
@dotnet-policy-service agree company="Duende Software"

halter73 commented on src/Security/Authentication/OpenIdConnect/src/OpenIdConnectHandler.cs:
```diff
@@ -485,6 +486,13 @@ private async Task HandleChallengeAsyncInternal(AuthenticationProperties propert
                 "Cannot redirect to the authorization endpoint, the configuration may be missing or invalid.");
         }
 
+        var parRequired = ConfigFlagEnabled("require_pushed_authoriation_requests");
```
```suggestion
        var parRequired = ConfigFlagEnabled("require_pushed_authorization_requests");
```

halter73 commented on src/Security/Authentication/OpenIdConnect/src/OpenIdConnectHandler.cs:
```diff
@@ -485,6 +486,13 @@ private async Task HandleChallengeAsyncInternal(AuthenticationProperties propert
 
+        var parRequired = ConfigFlagEnabled("require_pushed_authoriation_requests");
+
+        if (Options.UsePushedAuthorization || parRequired)
```
Do we need a way for the developer to say use pushed authorization if possible, but don't fail if there's no "pushed_authorization_request_endpoint" available?

Tratcher replied:
Do we expect this support to be variable? E.g. an auth server will support it or not, you should be able to determine that during development. If that support changes later I don't know that we should react dynamically in prod to it as there may be side-effects.

halter73 replied:
The change in behavior seems small. It doesn't even seem to prevent any callbacks from being called. It does change the parameters used for the user agent's ultimate authorization request, but I don't see how that's likely to cause issues. I'd be more worried about the OIDC provider requiring a custom authentication scheme to use the PAR endpoint, but we could fall back to not using PAR if given a non-200 response if PAR wasn't explicitly opted into. We could even stop trying PAR if the first attempt fails.

I think it's very possible that an OIDC provider that didn't support PAR during initial development might add support later. I understand not wanting to risk any potential breaks caused by an OIDC provider lighting up a feature that's not actually required for us to use, but it feels like we should default to the more secure option if it's available and it's unlikely to break anyone.

@blowdart do you have an opinion on this?

Tratcher replied:
An apps behavior should be in control of the developer, not a 3rd party, and shouldn't change without the developer instigating and testing that change.

halter73 replied:
The `OpenIdConnectHandler` lighting up features only if the `.well-known/openid-configuration` indicates the OIDC provider supports it is not breaking new ground. If you set `GetClaimsFromUserInfoEndpoint` to true, we will ignore it (aside from logging a debug-level message) if there is no "userinfo_endpoint".

https://github.com/dotnet/aspnetcore/blob/53db4d97d7c77d13e20e58a98f104e88d6af6040/src/Security/Authentication/OpenIdConnect/src/OpenIdConnectHandler.cs#L934-L940

We can debate what the default should be, but I think it's reasonable to want to have a way to say if the OIDC provider supports a feature that should enhance security, use it. I could see somebody wanting to do this in a library or code analyzer that ensures PAR is used when available.

josephdecock replied:
One thing that I've encountered is that some providers actually support PAR, but not for all clients. We ran into this when we added similar logic to IdentityModel.OidcClient (https://github.com/IdentityModel/IdentityModel.OidcClient/issues/428).

halter73 reviewed (commented):
Thanks for the PR. This helps a lot with the API review. I don't think I would have noticed the interaction between `UsePushedAuthorization` and "require_pushed_authorization_requests" without looking at the code for example.

Sorry for the delay reviewing the corresponding API proposal. Feel free to start on the tests. I think those will be easy to update even if we adjust the API.

halter73 commented on src/Security/Authentication/OpenIdConnect/src/Events/PushedAuthorizationContext.cs:
```diff
@@ -0,0 +1,59 @@
+    /// Initializes a new instance of <see cref="PushedAuthorizationContext"/>.
+    /// </summary>
+    /// <inheritdoc />
+    public PushedAuthorizationContext(HttpContext context, AuthenticationScheme scheme, OpenIdConnectOptions options, OpenIdConnectMessage message, HttpRequestMessage parRequest, AuthenticationProperties properties)
```
```suggestion
    public PushedAuthorizationContext(HttpContext context, AuthenticationScheme scheme, OpenIdConnectOptions options, OpenIdConnectMessage message, AuthenticationProperties properties)
```

If we think the `PushedAuthorizationContext` needs the `HttpRequestMessage` for some scenario, we should add a property for it instead.

josephdecock replied:
I've rearranged this, so we don't need a separate parameter at all. It turned out to be quite inconvenient to deal with the request message, because both the handler and the event might want to make changes to the parameters that will be sent.

halter73 commented:
@Tratcher Do you mind taking a look at this too? Thanks.

Tratcher commented on src/Security/Authentication/OpenIdConnect/src/OpenIdConnectOptions.cs:
```diff
@@ -408,4 +408,11 @@ public bool MapInboundClaims
     public bool UseSecurityTokenValidator { get; set; }
+
+    /// <summary>
+    /// Flag to set whether the handler should push authorization parameters on the backchannel before
```
Include an RFC link (or just #) for this scenario.

halter73 commented on src/Security/Authentication/OpenIdConnect/src/OpenIdConnectHandler.cs:
```diff
@@ -516,6 +524,113 @@ private async Task HandleChallengeAsyncInternal(AuthenticationProperties propert
+        var credential = $"{escapedClientId}:{escapedClientSecret}";
+        var encodedCredential = Convert.ToBase64String(Encoding.UTF8.GetBytes(credential));
+
+        parRequest.Headers.Authorization = new AuthenticationHeaderValue("basic", encodedCredential);
```
Why was this header chosen for authentication by default? Looking at [RFC 9126](https://datatracker.ietf.org/doc/html/rfc9126#section-2), it allows authenticating the same way you would with the token endpoint as defined in [RFC 6749](https://datatracker.ietf.org/doc/html/rfc6749#section-3.2.1). 

I don't see anything where the client ID and client secret are put into a basic auth header. The "introductory example" in RFC 9126 uses "JWT client assertion-based authentication" defined in [RFC 7523](https://datatracker.ietf.org/doc/html/rfc7523#section-2.2), but that doesn't seem to provide a mechanism to acquire the JWT.

The common thing for the token endpoint seems to be to put the client ID and client secret in the request body of the token endpoint even though it's "NOT RECOMMENDED and SHOULD be limited to clients unable to directly utilize the HTTP Basic authentication scheme."

Why not just include the client secret in the `parRequest` body via `authorizeRequest.Parameters` since it's never seen by the client? It already includes the client ID, the spec seems to allow it, and it's effectively what we do by default for the token endpoint, right?

> The request also includes, as appropriate for the given client, any additional parameters necessary for client authentication (e.g., client_secret or client_assertion and client_assertion_type)

https://datatracker.ietf.org/doc/html/rfc9126#section-2.1

josephdecock replied:
I think you're right - I'll switch to the request body style.

I was trying to follow the "NOT RECOMMENDED" language, despite the inconsistency with the way we currently call the token endpoint. I had some thought that we might want to change _that_ as well, but the latest draft of OAuth 2.1 actually [drops ](https://author-tools.ietf.org/iddiff?url1=draft-ietf-oauth-v2-1-10&url2=draft-ietf-oauth-v2-1-11&difftype=--html) this language:

> Change client credentials to be required to be supported in the request body to avoid HTTP Basic authentication encoding interop issues

So that really makes it a moot point, and we should definitely just use the request body.

blowdart replied:
Having basic auth of any description would be rather problematic from a security POV :)

halter73 commented on src/Security/Authentication/OpenIdConnect/src/OpenIdConnectHandler.cs:
```diff
@@ -516,6 +524,113 @@ private async Task HandleChallengeAsyncInternal(AuthenticationProperties propert
+                Content = new FormUrlEncodedContent(authorizeRequest.Parameters),
+            };
+            var context = new PushedAuthorizationContext(Context, Scheme, Options, authorizeRequest, parRequest, properties);
+            await Events.PushAuthorization(context);
```
I just noticed that the `authorizeRequest.Parameters` are already serialized into the `FormUrlEncodedContent` by the time this event is called. I know that's probably because you want to allow the event to modify the `HttpRequestMessage`, but I think we should set `HttpRequestMessage.Content` after this event if it's still null.

josephdecock replied:
Instead of allowing the event to modify the request message, I'm just passing an `OpenIdConnectMessage` that the handler later pushes. With no message, we don't have to worry about the message content being null.

The context now can affect the flow of control in the handler in 3 ways:
- If it handles authentication with `HandleClientAuthentication`, we skip setting the client secret
- If it decides that we don't need to push with `SkipPush`, we completely skip pushing, and redirect as normal
- If it decides to completely control the call to the par endpoint with `HandlePush`, the event won't make the par request and instead will use the `request_uri ` passed to `HandlePush`
- The event can also tweak parameters before they are pushed, but not call any of the context's methods to get otherwise default behavior.

The downside to this is that it is possible to use the API in an ambiguous way: what should the handler do if the event calls both `HandlePush` and `SkipPush`? Should that be an error? A logged warning? Should the last call take precedence?

halter73 replied:
> The downside to this is that it is possible to use the API in an ambiguous way: what should the handler do if the event calls both `HandlePush` and `SkipPush`? Should that be an error? A logged warning? Should the last call take precedence?

I think throwing an `InvalidOperationException` is a good choice if we detect this. I don't think it should depend on order.

halter73 commented on src/Security/Authentication/OpenIdConnect/src/OpenIdConnectHandler.cs:
```diff
@@ -516,6 +524,113 @@ private async Task HandleChallengeAsyncInternal(AuthenticationProperties propert
+        }
+    }
+
+    private async Task<string> GetRequestUri(HttpResponseMessage parResponseMessage)
```
Nit:

```suggestion
    private async Task<string> GetPushAuthorizationRequestUri(HttpResponseMessage parResponseMessage)
```

blowdart commented:
_I think it's very possible that an OIDC provider that didn't support PAR during initial development might add support later. I understand not wanting to risk any potential breaks caused by an OIDC provider lighting up a feature that's not actually required for us to use, but it feels like we should default to the more secure option if it's available and it's unlikely to break anyone._

We should be more secure by default, and if that breaks people give them an opt-out

josephdecock committed:
PAR updates from review

josephdecock committed:
Simplify PAR event context

josephdecock committed:
Clarify a few PAR related comments

josephdecock committed:
Enable PAR by default, if available in disco

josephdecock committed:
Very minor changes for clarity

halter73 commented:
Just to give you a heads up, code complete for preview 6 is this Wednesday. I know that's very soon, so I'm fine with waiting for preview 7 which is scheduled to be code complete July 19th. I think that's the latest we'd take this change.

halter73 commented:
Remember, code complete for the last preview is this Friday, July 19th.

josephdecock committed:
Add enum to control PAR behavior

josephdecock committed:
Added basic tests for pushed authorization

josephdecock committed:
Seal PushedAuthorizationContext

josephdecock committed:
Add more PAR unit tests

josephdecock committed:
Clean up compiler warnings

josephdecock committed:
Revert commented out config in sample

josephdecock committed:
Revert work in progress test

josephdecock committed:
Clean up test

josephdecock committed:
Disallow multiple PAR context actions

You can indicate if the event handled authentication, handled the push entirely, or opted out of pushing. Doing more than 1 of those is at best redundant but often ambiguous, so it is always disallowed.

josephdecock commented:
@halter73 I marked this as ready for review last night, then woke up and realized I wanted to do a bit more cleanup and handle the error case of the event calling multiple context methods (e.g., you shouldn't be able to both say that pushing is skipped, but also you handled the authentication part of the push). That's all in now, and it is really, truly, for-real-this-time ready for review.

Thanks!

halter73 commented on src/Security/Authentication/OpenIdConnect/src/OpenIdConnectHandler.cs:
```diff
@@ -516,6 +552,113 @@ private async Task HandleChallengeAsyncInternal(AuthenticationProperties propert
     }
 
+    // TODO GetConfigString and ConfigFlagEnabled are only used in PAR, and won't be necessary after
+    // Microsoft.IdentityModel.Protocols.OpenIdConnect is updated to v7.6.1 or later.
```
We've updated to 8.0.0, so we should be able to use `PushedAuthorizationRequestEndpoint` and the like now.

josephdecock replied:
Awesome, I switched to it!

josephdecock replied:
Out of curiosity, how are the dependencies managed? The build system is ... well it's a lot for the uninitiated 😄 

I found eng/BaselineDesigner.props, but it has this:
```
  <ItemGroup Condition=" '$(PackageId)' == 'Microsoft.AspNetCore.Authentication.JwtBearer' AND ('$(TargetFramework)' == '$(DefaultNetCoreTargetFramework)' OR '$(TargetFramework)' == 'net8.0') ">
    <BaselinePackageReference Include="Microsoft.IdentityModel.Protocols.OpenIdConnect" Version="[7.0.3, )" />
  </ItemGroup>
```
That looks like 7.0.3 is being referenced, which is clearly not the case.

halter73 replied:
The dependency versions we use come from [Version.props](https://github.com/dotnet/aspnetcore/blob/dfc5189cb5b9ecfd0e0c43310a9355e768bbed45/eng/Versions.props#L14). BaselineDesigner.props is to catch unintentional breaking changes in servicing builds. This should provide some more context.

https://github.com/dotnet/aspnetcore/blob/dfc5189cb5b9ecfd0e0c43310a9355e768bbed45/eng/targets/ResolveReferences.targets#L11-L17

halter73 commented on src/Security/Authentication/OpenIdConnect/src/PushedAuthorizationBehavior.cs:
```diff
@@ -0,0 +1,28 @@
+    UseIfAvailable,
+    /// <summary>
+    /// Never use Pushed Authorization (PAR), even if the PAR endpoint is available in the identity provider's discovery document or the explicit <see cref="OpenIdConnectConfiguration"/>.
+    ///
+    /// TODO - Document the required by disco but disabled here case
+    ///
```
Does this sound reasonable?

```suggestion
    /// If the identity provider's discovery document indicates that it requires Pushed Authorization (PAR), the handler will fail.
```

halter73 commented on src/Security/Authentication/OpenIdConnect/src/OpenIdConnectHandler.cs:
```diff
@@ -516,6 +552,113 @@ private async Task HandleChallengeAsyncInternal(AuthenticationProperties propert
+        }
+
+        // Parse response
+        var parResponseString = await parResponseMessage.Content.ReadAsStringAsync();
```
```suggestion
        var parResponseString = await parResponseMessage.Content.ReadAsStringAsync(Context.RequestAborted);
```

halter73 commented on src/Security/Authentication/OpenIdConnect/src/OpenIdConnectHandler.cs:
```diff
@@ -516,6 +552,113 @@ private async Task HandleChallengeAsyncInternal(AuthenticationProperties propert
+        authorizeRequest.Parameters.Add("request_uri", requestUri);
+    }
+
+    // TODO Compare with similar use case in RedeemAuthorizationCodeAsync
```
```suggestion
```

Is there anything that `RedeemAuthorizationCodeAsync` that we'd still want to copy. I see that it's less picky about the MediaType, but I like being stricter.

josephdecock replied:
I agree, we're good

josephdecock committed:
Update src/Security/Authentication/OpenIdConnect/src/OpenIdConnectHandler.cs

Co-authored-by: Stephen Halter <halter73@gmail.com>

josephdecock committed:
Update src/Security/Authentication/OpenIdConnect/src/PushedAuthorizationBehavior.cs

Co-authored-by: Stephen Halter <halter73@gmail.com>

josephdecock committed:
Update src/Security/Authentication/OpenIdConnect/src/OpenIdConnectHandler.cs

Co-authored-by: Stephen Halter <halter73@gmail.com>

josephdecock committed:
Fix exception messages when multiple PAR event methods are called

josephdecock committed:
Merge branch 'josephdecock/par' of https://github.com/josephdecock/aspnetcore into josephdecock/par

josephdecock commented on src/Security/Authentication/OpenIdConnect/src/Events/PushedAuthorizationContext.cs:
```diff
@@ -0,0 +1,106 @@
+    {
+        if (SkippedPush || HandledClientAuthentication)
+        {
+            throw new InvalidOperationException("TODO MESSASGE");
```
Woops, thanks for catching that...

Commit:
Merge branch 'main' into josephdecock/par

josephdecock committed:
Use PAR support from System.IdentityModel to simplify discovery

josephdecock commented:
@javiercn could you please retry the pipeline failures?

halter73 reviewed (approved):
Thank you!

halter73 commented on src/Security/Authentication/OpenIdConnect/src/OpenIdConnectHandler.cs:
```diff
@@ -516,6 +550,82 @@ private async Task HandleChallengeAsyncInternal(AuthenticationProperties propert
+            var requestMessage = new HttpRequestMessage(HttpMethod.Post, parEndpoint);
+            requestMessage.Content = new FormUrlEncodedContent(parRequest.Parameters);
+            requestMessage.Version = Backchannel.DefaultRequestVersion;
+            var parResponseMessage = await Backchannel.SendAsync(requestMessage, Context.RequestAborted);
```
> One thing that I've encountered is that some providers actually support PAR, but not for all clients. We ran into this when we added similar logic to IdentityModel.OidcClient (https://github.com/IdentityModel/IdentityModel.OidcClient/issues/428).

I just noticed this from the previous thread about making `UseIfAvailable` the default. I was also concerned that there may be providers that authenticate the PAR request with something other than the client_secret parameter. Would it make sense to log a warning and fallback to a normal non-pushed authorization attempt if we cannot get a PAR request URI in the `UseIfAvailable` case?

I'm still fine with merging this in a preview to see how it goes, but not many people use previews compared to major releases. It might be a good idea for us to update this behavior in RC1 even if no one complains to be extra safe, but I'm curious what everyone else thinks about this.

josephdecock replied:
On the one hand, in the issue I linked to in IdentityModel.OidcClient, the issue was for public clients specifically and most users of the OidcHandler will be confidential clients. 

On the other hand, if you have customized the behavior of the handler to call the token endpoint using a different form of authentication, presumably you need to do the same for the PAR requests that UseIfAvailable will start making for you. 

But, users who are customizing the handler's authentication already are advanced users and personally I'm okay with expecting them to sort this out.

My feeling is to keep the UseIfAvailable default. For RC1, we could take a look at exactly what the error message is if authentication to the PAR endpoint fails and log a hint about the new APIs.
